core/src/tx.rs

```diff
@@ -14,19 +14,129 @@
+}
+
+impl AcceptedTransaction {
+    /// Accept genesis transaction. Transition from [`GenesisTransaction`] to [`AcceptedTransaction`].
+    pub fn accept_genesis(tx: GenesisTransaction) -> Self {
+        Self(tx.0)
+    }
+
+    /// Accept transaction. Transition from [`VersionedSignedTransaction`] to [`AcceptedTransaction`].
+    ///
+    /// # Errors
+    ///
+    /// - if it does not adhere to limits
+    pub fn accept(
+        transaction: VersionedSignedTransaction,
+        limits: &TransactionLimits,
+    ) -> Result<Self, AcceptTransactionFail> {
+        match &transaction.payload().instructions {
+            Executable::Instructions(instructions) => {
+                let instruction_count: u64 = instructions
+                    .iter()
+                    .map(InstructionBox::len)
+                    .sum::<usize>()
+                    .try_into()
+                    .expect("`usize` should always fit in `u64`");
+
+                if instruction_count > limits.max_instruction_number {
+                    return Err(AcceptTransactionFail::TransactionLimit(
+                        TransactionLimitError {
+                            reason: format!(
+                                "Too many instructions in payload, max number is {}, but got {}",
+                                limits.max_instruction_number, instruction_count
+                            ),
+                        },
+                    ));
+                }
+            }
+            // TODO: Can we check the number of instructions in wasm? Because we do this check
+            // when executing wasm where we deny wasm if number of instructions exceeds the limit.
+            //
+            // Should we allow infinite instructions in wasm? And deny only based on fuel and size
+            Executable::Wasm(smart_contract) => {
+                let max_wasm_size_bytes = limits.max_wasm_size_bytes;
+
+                let size_bytes: u64 = smart_contract
+                    .size_bytes()
+                    .try_into()
+                    .expect("`u64` should always fit in `u64`");
+
+                if size_bytes > max_wasm_size_bytes {
+                    return Err(AcceptTransactionFail::TransactionLimit(
+                        TransactionLimitError {
+                            reason: format!("Wasm binary too large, max size is {max_wasm_size_bytes}, but got {size_bytes}"),
+                        },
+                    ));
+                }
+            }
+        }
+
+        Ok(Self(transaction))
+    }
+
+    /// Transaction hash
+    pub fn hash(&self) -> HashOf<VersionedSignedTransaction> {
+        self.0.hash()
+    }
+
+    /// Payload of the transaction
+    pub fn payload(&self) -> &TransactionPayload {
+        self.0.payload()
+    }
+
+    pub(crate) fn signatures(&self) -> &SignaturesOf<TransactionPayload> {
+        self.0.signatures()
+    }
```

Comment:
Maybe implement `Deref`?

Reply:
I could. But semantically it would be more appropriate to implement `AsRef`

Reply:
I mean that we `Deref` we could just call `tx.payload()` without implementing this methods for `VersionedSignedTransaction`.

Reply:
After reading attached post i agree with your point.